alpha/serial_cmd.c

```diff
@@ -0,0 +1,162 @@
+ CONDITIONS OF ANY KIND, either express or implied.
+*/
+
+#include <stdio.h>
+#include "esp_console.h"
+#include "esp_vfs_dev.h"
+#include "driver/uart.h"
+#include "linenoise/linenoise.h"
+#include "argtable3/argtable3.h"
+#include "serial_cmd.h"
+
+static bool uart_initialized = false;
+
+static void register_version();
+
+void register_commands()
+{
+  register_version();
+}
+
+void initialize_uart()
+{
+  if (uart_initialized == true)
+    return;
+
+  /* Configure UART. Note that REF_TICK is used so that the baud rate remains
+   * correct while APB frequency is changing in light sleep mode.
+   */
+  const uart_config_t uart_config = {
+          .baud_rate = 115200,
+          .data_bits = UART_DATA_8_BITS,
+          .parity = UART_PARITY_DISABLE,
+          .stop_bits = UART_STOP_BITS_1,
+          .use_ref_tick = true
+  };
+  ESP_ERROR_CHECK( uart_param_config(CONFIG_ESP_CONSOLE_UART_NUM, &uart_config) );
+
+  /* Install UART driver for interrupt-driven reads and writes */
+  ESP_ERROR_CHECK( uart_driver_install(CONFIG_ESP_CONSOLE_UART_NUM,
+          256, 0, 0, NULL, 0) );
+
+  uart_initialized = true;
+}
+
+void initialize_console()
+{
+  /* Disable buffering on stdin */
+  setvbuf(stdin, NULL, _IONBF, 0);
+
+  /* Minicom, screen, idf_monitor send CR when ENTER key is pressed */
+  esp_vfs_dev_uart_set_rx_line_endings(ESP_LINE_ENDINGS_CR);
+  /* Move the caret to the beginning of the next line on '\n' */
+  esp_vfs_dev_uart_set_tx_line_endings(ESP_LINE_ENDINGS_CRLF);
+
+  /* Tell VFS to use UART driver */
+  esp_vfs_dev_uart_use_driver(CONFIG_ESP_CONSOLE_UART_NUM);
+
+  /* Initialize the console */
+  esp_console_config_t console_config = {
+          .max_cmdline_args = 8,
+          .max_cmdline_length = 256,
+  };
+  ESP_ERROR_CHECK( esp_console_init(&console_config) );
+
+  /* Configure linenoise line completion library */
+  /* Enable multiline editing. If not set, long commands will scroll within
+   * single line.
+   */
+  linenoiseSetMultiLine(1);
+
+  /* Tell linenoise where to get command completions and hints */
+  linenoiseSetCompletionCallback(&esp_console_get_completion);
+  linenoiseSetHintsCallback((linenoiseHintsCallback*) &esp_console_get_hint);
+
+  /* Set command history size */
+  linenoiseHistorySetMaxLen(100);
+
+#if CONFIG_STORE_HISTORY
+  /* Load command history from filesystem */
+  linenoiseHistoryLoad(HISTORY_PATH);
+#endif
+}
+
+void serial_cmd_loop()
+{
+  initialize_uart();
+  initialize_console();
+  register_commands();
+
+  /* Prompt to be printed before each line.
+   * This can be customized, made dynamic, etc.
+   */
+  const char* prompt = "esp32> ";
+  linenoiseSetDumbMode(1);
+
+  printf("\n"
+         "This is an example of ESP-IDF console component.\n"
+         "Type 'help' to get the list of commands.\n"
+         "Use UP/DOWN arrows to navigate through command history.\n"
+         "Press TAB when typing command name to auto-complete.\n");
+
+  /* Main loop */
+  while(true) {
```

Comment:
Blocking loop but serial is just for onboarding/initial setup right?

Reply:
Yup that's right